File: 01-hard/as_quick_as_a_flash/as.quick.as.a.flash.cpp
// Reuben Thorpe (2016), CodeEval [As Quick As A Flash v1.1]
#include <iostream>
#include <sstream>
#include <vector>
#include <fstream>

using namespace std;

int q_sort_pivots(vector<int> seq, int begin, int end);
void swap(int& a, int& b);


int main(int argc, char* argv[]) {
  ifstream inFile(argv[1]);
  vector<int> seq;
  seq.reserve(30); // Maximum number of elements
  int num, N;
  string line;

  while (getline(inFile, line)) {
    stringstream ss;
    ss.str(line);
    while (ss >> num) seq.push_back(num);
    N = seq.size() - 1;
    cout << q_sort_pivots(seq, 0, N) << "\n";
    seq.clear();
  }

  return(0);
}


int q_sort_pivots(vector<int> seq, int begin, int end) {
  if (begin < end) {
    int pivot = seq[begin];
    int pivot_pos = begin;
    int left = begin+1;
    int right = end;

    while (left < right) {
      while (left < right && seq[right] >= pivot) right--;

      if (seq[right] < pivot) {
        swap(seq[pivot_pos], seq[right]);
        pivot_pos = right;
      }

      if (left < right) {
        while (left < right && seq[left] <= pivot) left++;

        if (seq[left] > pivot) {
          swap(seq[pivot_pos], seq[left]);
          pivot_pos = left;
        }
      }
    }

    return(1 + q_sort_pivots(seq, begin, pivot_pos-1) +
           q_sort_pivots(seq, pivot_pos+1, end));
  }
  else return(0);
}


void swap(int& a, int& b) {
  int tmp = a;
  a = b;
  b = tmp;
}

File: code.eval/01-hard/as_quick_as_a_flash/as.quick.as.a.flash.py
# Reuben Thorpe (2016), CodeEval [As Quick As A Flash v1.0]
from sys import argv


def swap(seq, x, y):
    seq[x], seq[y] = seq[y], seq[x]


def q_sort_pivots(seq, begin, end):
    """
        Code Evals implimentation of the quick sort algorithm, returns the
        number of pivots enacted during a complete sort.
    """

    if begin < end:
        pivot = seq[begin]
        pivot_pos = begin
        left = begin+1
        right = end

        while left < right:
            while (left < right and seq[right] >= pivot):
                right -= 1

            if seq[right] < pivot:
                swap(seq, pivot_pos, right)
                pivot_pos = right

            if left < right:
                while (left < right and seq[left] <= pivot):
                    left += 1

                if seq[left] > pivot:
                    swap(seq, pivot_pos, left)
                    pivot_pos = left

        return(1 + q_sort_pivots(seq, begin, pivot_pos-1) +
               q_sort_pivots(seq, pivot_pos+1, end))

    else:
        return(0)


if __name__ == "__main__":
    for seq in open(argv[1], "r"):
        seq = [int(num) for num in seq.split()]
        N = len(seq) - 1
        print(q_sort_pivots(seq, 0, N))
